corehq/apps/es/README.rst

```diff
@@ -0,0 +1,220 @@
+Elasticsearch App
+=================
+
+Adapter Design
+--------------
+
+The HQ Elastic adapter design came about due to the need for reindexing
+Elasticsearch indexes in a way that is transparent to parts of HQ that write to
+Elasticsearch (e.g. pillowtop).  Reindexing is necessary for making changes to
+index mappings, is a prerequisite to upgrading an Elasticsearch cluster, and is
+also needed for changing low-level index configurations (e.g. sharding).
+
+There is an existing procedure draft that documents the steps that were used on
+one occasion to reindex the ``case_search`` index.  This procedure leveraged a
+custom pillow to "backfill" the cloned index (i.e. initially populated using
+Elasticsearch Reindex API). That procedure only works for a subset of HQ
+Elasticsearch indexes, and is too risky to be considered as an ongoing Elastic
+maintenance strategy. There are several key constraints that an HQ reindexing
+procedure should meet which the existing procedure does not:
+
+- simple and robust
+- performed with standard maintenance practices
+- provides the ability to test and verify the integrity of a new index before it
+  is too late to be rejected
+- allows HQ Elasticsearch index state to remain decoupled from the
+  commcare-cloud codebase
+- is not disruptive -- does not prohibit any other kind of standard maintenance
+  that might come up while the operation is underway
+- is "fire and forget" -- does not require active polling of intermediate state
+  in order to progress the overall operation
+- is practical for third party HQ hosters to use
```

Comment:
@joxl It would be great to understand more about how these goals will be accomplished.

Reply:
> 3rd parties will be forced to reindex in the timeframes that we dictate.

This is not new. 3rd parties are already forced to reindex on our schedule. All but one or two Elasticsearch index names (which dictates a reindex) live in pillowtop code. 3rd parties have no control over this.

> to create the new index and then again to remove the new index.

New index creation does not require a manual step -- the multiplexing adapter will create the index when it starts up (just like pillows does now). The steps are 1) idempotent "sync" command and 2) delete the old index.  The current reindex process has these same "functional" steps, only the first is currently a post-deploy operation. The second (delete the old index) is currently a manual step AFAIK, so that's not new.

> Will that also be required for 3rd party hosters?

Yes, the same changelog process we use now to guide 3rd party hostsers through reindex operations will be used for multiplexing reindex operations.  There will also be safeguards in place to prevent accidentally missing steps and putting the Elastic cluster in an inconsistent state. Current reindex procedures do not have these sort of automatic safety measures.